String() fails on map[string]string

```go
package main

import (
	"fmt"

	toml "github.com/pelletier/go-toml"
)

func main() {
	m1 := map[string]string{"v": "abc"}
	m2 := map[string]interface{}{"m": m1}
	tree := toml.TreeFromMap(m2)
	fmt.Println(tree.String())
}
```

Panics:

```
panic: unsupported value type: map[v:abc]
```

In Hugo we do frontmatter conversions between YAML and TOML etc., and the datatypes for maps differs slightly. We recently converted from Burntsushi to pelletier's go-toml for Hugo and related libraries. We are in the process to release a new version of Hugo and need to fix this in a way or another.

See https://github.com/spf13/hugo/issues/2439

--- a/parser_test.go
+++ b/parser_test.go
@@ -658,6 +658,28 @@ func TestToString(t *testing.T) {
 	}
 }
 
+func TestToStringMapStringString(t *testing.T) {
+	in := map[string]interface{}{"m": map[string]string{"v": "abc"}}
+	want := "\n[m]\n  v = \"abc\"\n"
+	tree := TreeFromMap(in)
+	got := tree.String()
+
+	if got != want {
+		t.Errorf("want:\n%q\ngot:\n%q", want, got)
+	}
+}
+
+func TestToStringMapInterfaceInterface(t *testing.T) {
+	in := map[string]interface{}{"m": map[interface{}]interface{}{"v": "abc"}}
+	want := "\n[m]\n  v = \"abc\"\n"
+	tree := TreeFromMap(in)
+	got := tree.String()
+
+	if got != want {
+		t.Errorf("want:\n%q\ngot:\n%q", want, got)
+	}
+}
+
 func assertPosition(t *testing.T, text string, ref map[string]Position) {
 	tree, err := Load(text)
 	if err != nil {

--- a/parser_test.go
+++ b/parser_test.go
@@ -637,6 +637,7 @@ func TestToTomlValue(t *testing.T) {
 			"1979-05-27T07:32:00Z"},
 		{[]interface{}{"gamma", "delta"},
 			"[\n  \"gamma\",\n  \"delta\",\n]"},
+		{nil, ""},
 	} {
 		result := toTomlValue(item.Value, 0)
 		if result != item.Expect {

--- a/tomltree_conversions.go
+++ b/tomltree_conversions.go
@@ -100,11 +100,41 @@ func (t *TomlTree) toToml(indent, keyspace string) string {
 				result += fmt.Sprintf("\n%s[%s]\n", indent, combinedKey)
 			}
 			result += sub.toToml(indent+"  ", combinedKey)
+		case map[string]string:
+			sub := TreeFromMap(convertMapStringString(node))
+
+			if len(sub.Keys()) > 0 {
+				result += fmt.Sprintf("\n%s[%s]\n", indent, combinedKey)
+			}
+			result += sub.toToml(indent+"  ", combinedKey)
+		case map[interface{}]interface{}:
+			sub := TreeFromMap(convertMapInterfaceInterface(node))
+
+			if len(sub.Keys()) > 0 {
+				result += fmt.Sprintf("\n%s[%s]\n", indent, combinedKey)
+			}
+			result += sub.toToml(indent+"  ", combinedKey)
 		case *tomlValue:
 			result += fmt.Sprintf("%s%s = %s\n", indent, k, toTomlValue(node.value, 0))
 		default:
 			result += fmt.Sprintf("%s%s = %s\n", indent, k, toTomlValue(v, 0))
 		}
+	}
+	return result
+}
+
+func convertMapStringString(in map[string]string) map[string]interface{} {
+	result := make(map[string]interface{}, len(in))
+	for k, v := range in {
+		result[k] = v
+	}
+	return result
+}
+
+func convertMapInterfaceInterface(in map[interface{}]interface{}) map[string]interface{} {
+	result := make(map[string]interface{}, len(in))
+	for k, v := range in {
+		result[k.(string)] = v
 	}
 	return result
 }

--- a/tomltree_conversions.go
+++ b/tomltree_conversions.go
@@ -64,6 +64,8 @@ func toTomlValue(item interface{}, indent int) string {
 			result += toTomlValue(item, indent+2) + ",\n"
 		}
 		return result + tab + "]"
+	case nil:
+		return ""
 	default:
 		panic(fmt.Sprintf("unsupported value type: %v", value))
 	}

